### ISSUE TITLE
Insert or Update

### ISSUE THREAD

**Issue by eugenkiss** [requery/requery#10]
Hey,

is there a built-in insert or update functionality that I missed? If there does not exist one yet what would you propose as a good workaround?

For now, I have implemented the following specialised extension function in Kotlin (corresponding Java code should be hopefully clear).

```kotlin
fun <E: Persistable> SingleEntityStore<Persistable>.upsert(entity: E): Single<E> {
    return insert(entity).toObservable().onErrorResumeNext { update(entity).toObservable() }.toSingle()
}
```

~~It seems to work~~ Not really, I just noticed. Need to improve it. 

I tried to create a more generic version.

```kotlin
fun <T, R, E : T> EntityStore<T, R>.upsert(entity: E): R {
    try {
        return insert(entity)
    } catch (e: PersistenceException) {
        return update(entity)
    }
}
```

Of course, this won't work as the `insert`/`update` operations are executed at a later point in time.

In any case, I'd vote for a native upsert implementation ;).

**Comment by npurushe** [comment]
Hi yes upsert is a good thing. Unfortunately the databases that support it all do it a different way so tricky to implement.

Is this for SQLite? SQLite only has `insert or replace` which is not the same as upsert unfortunately. I think for SQLite if you have some primary key it's best to check if that key already has a record first and retrieve it and base your insert/update decision on that, while of course being careful of concurrency.

**Comment by eugenkiss** [comment]
Yes, it's for SQLite in Android. Thanks for your tips. I've create two new versions of upsert (or rather a rough approximations thereof as you correctly mentioned that SQLite does not really have upsert support) extension functions.

```kotlin
fun <E: Persistable, K> SingleEntityStore<Persistable>.upsert(clazz: Class<E>, key: K, entity: E): Single<E> {
    return findByKey(clazz, key).toObservable().materialize().flatMap { notification ->
        if (notification.isOnError) {
            insert(entity).toObservable()
        } else {
            update(entity).toObservable()
        }
    }.toSingle()
}
```

```kotlin
fun <E: Persistable, K> BlockingEntityStore<Persistable>.upsert(clazz: Class<E>, key: K, entity: E): E {
    try {
        findByKey(clazz, key) == null
    } catch (e: NoSuchElementException) {
        return insert(entity)
    }
    return update(entity)
}
```

With the first one I had some concurrency problems (where I upserted a list on entities). I wasn't patient enough to look into fixing them so I created an extension function for a blocking version which now seems to work alright. 

By the way, there seems to still be something fishy when I look at Requery's `beforeExecuteQuery` logs. Basically, I have five entities that I upsert and there are four distinct entities in the database. The first and the last of the five entities represent the same one in the database. The logs always show only *four* select statements but five insert or update statements. Can it be that a the last select is not executed because the entity is found in the cache? If yes, maybe that should be indicated in the logs?
